packager/packager.cc

```diff
@@ -410,26 +416,38 @@ class PackageInfoImpl final : public core::packages::PackageInfo {
     }
 
     optional<core::AutocorrectSuggestion> addImport(const core::GlobalState &gs, const PackageInfo &pkg,
-                                                    bool isTestImport) const {
+                                                    core::packages::ImportType importType) const {
         auto &info = PackageInfoImpl::from(pkg);
         auto insertionLoc = core::Loc::none(loc.file());
         optional<core::AutocorrectSuggestion::Edit> deleteTestImportEdit = nullopt;
         if (!importedPackageNames.empty()) {
             core::LocOffsets importToInsertAfter;
             for (auto &import : importedPackageNames) {
                 if (import.name.mangledName == info.name.mangledName) {
-                    if (!isTestImport && import.type == core::packages::ImportType::Test) {
-                        // There's already a test import for this package, so we'll convert it to a regular import.
-                        // importToInsertAfter already tracks where we need to insert the import.
-                        // So we can craft an edit to delete the `test_import` line, and then use the regular logic for
-                        // adding an import to insert the `import`.
-                        auto importLoc = core::Loc(fullLoc().file(), import.name.fullName.loc);
+                    if ((importType == core::packages::ImportType::Normal &&
+                         import.type != core::packages::ImportType::Normal) ||
+                        (importType == core::packages::ImportType::TestHelper &&
+                         import.type == core::packages::ImportType::TestUnit)) {
+                        // There's already an import for this package, so we'll "upgrade" it to the desired import.
+                        // importToInsertAfter already tracks where we need to insert the import.  So we can craft an
+                        // edit to delete the existing line, and then use the regular logic for adding an import to
+                        // insert the `import`.
+                        auto importLoc = core::Loc(fullLoc().file(), import.loc);
                         auto [lineStart, numWhitespace] = importLoc.findStartOfIndentation(gs);
                         auto beginPos =
                             lineStart.adjust(gs, -numWhitespace, 0).beginPos(); // -numWhitespace for the indentation
                         auto endPos = importLoc.endPos();
                         core::Loc replaceLoc(importLoc.file(), beginPos, endPos);
+
                         deleteTestImportEdit = {replaceLoc, ""};
+
+                        // as a special-case: if we're converting a `test_import` to remove `only:`, then we want to
+                        // re-insert it at this exact same point (since we sort those together.) Let's find the previous
+                        // import!
```

Comment:
I don't think you need to do it now (so feel free to add a TODO for it), but it might be nice to create an autocorrect that directly removes the `, only:`, instead of deleting the entire line and adding a new import line. It's more special case logic, but will be less noisy in the autocorrect output.